mitmproxy/proxy/layers/quic.py

```diff
@@ -1257,7 +1290,13 @@ def receive_handshake_data(
             return False, "connection closed early"
 
         # send the client hello to aioquic
-        return (yield from super().receive_handshake_data(data))
+        ret: tuple[bool, str | None]
+        for msg in self.msgs:
+            ret = yield from super().receive_handshake_data(msg)
+            if not ret[0]:
+                break
```

Comment:
Why `break` here? Do we not want to send the remaining packets to aioquic?

(if there is a good reason for this: add an explanatory comment)

Reply:
> Do we not want to send the remaining packets to aioquic?

Yes, we want to close the connection if for any packet we receive `done=False` from `super().receive_handshake_data`. There's no QUIC-specific reason behind this but I don't see any other way to handle the case where say the second packet fails and we still have a third packet to be sent.

Reply:
Now I'm more confused than before - if we receive out-of-order packets, isn't `done=False` expected after the first packet?

Reply:
Oh yes makes sense, I think I've interpreted things incorrectly. But what's weirder is that HTTP3 (reverse proxy) works irrespective of the `break` statement. Digging deeper, it seems we always get `done=False` here because of the `quic_events.ProtocolNegotiated` event raised in `QuicLayer.receive_handshake_data` and we can't return `True, None` for this event yet since the handshake isn't completed yet.

Reply:
> But what's weird is that HTTP3 (reverse proxy) works irrespective of the break statement.

(this could also be due to retransmissions)

> it seems we always get `done=False` here because we raise `quic_events.ProtocolNegotiated` event in `QuicLayer.receive_handshake_data`

From a very cursory read, we swallow that immediately and continue with the. `while event := self.quic.next_event():` loop:
https://github.com/mitmproxy/mitmproxy/blob/0f42f00baa2925964213ed5c28aac1655abbe0db/mitmproxy/proxy/layers/quic.py#L954-L963

Without being fully in the weeds here: Intuitively we can maybe just do something like this?

```python
for msg in self.msgs:
    self.quic.receive_datagram(data, self.conn.peername, now=self._time())

return (yield from super().receive_handshake_data(b""))
```

Reply:
> Without being fully in the weeds here: Intuitively we can maybe just do something like this?

Yup, this works as well.